design/REGO_V2_PROPOSAL.md

```diff
@@ -184,44 +220,79 @@ cannot be reassigned.
+conflict resolution behavior among obligations, this can radically affect
+request behavior. Future revisions of this spec should address the different
+kinds of obligations that may be emitted and conflict resolution strategies to
+be employed during execution, but for now they are broken down into two
+categories: synchronous and asynchronous. Synchronous obligations may include
+preconditions, whereas async obligations would be considered promises. Advice,
+on the other hand, should always be considered asynchronous and best-effort.
+
+In the rule examples, the decision is binary, so the first argument to the
 trigger has been labelled allow although it could be any type of decision. The
 second argument is a dynamic object representing context supplied to the
-policies responsible for the decision.
+rules responsible for the decision.
 
 ```
-// Triggers are evaluated after policy decisions and have a void return.
-@trigger function onSalaryRequest(allow, ctx) {
-  // Triggers may inspect the policy signatures invoked and their outcomes.
-  if (ctx.request.method in ['get', 'list'] && !viewSalary) {
-    logger.log(ctx.user + " denied view salary request" + ctx.resource);
+// Triggers are evaluated after rule decisions and have a void return.
+@trigger function onSalaryRequest(decision, ctx) {
+  // Triggers may invoke rules without affecting the rule outcome.
+  if (ctx.request.method == 'list'
+      && !queryDepartmentSalaries(ctx.request, ctx.resource, ctx.user)) {
+    logger.log(ctx.user + " denied view salary request " + ctx.resource);
   }
 
   // Alternatively, triggers may simply inspect context to make a decision.
   match_result = ctx.resource.match('users/{target_user}/**');
   target_user = match_result.groups.target_user;
-  if (ctx.user != target_user && allow) {
+  if (ctx.user != target_user && decision == allow) {
     logger.log(ctx.user + " perform a salary action on " + target_user);
   }
 }
 ```
 
 Triggers are the only functions that can and must have a void return type. They
 must not be referenced within non-trigger functions, though they may invoke other
-non-policy functions including other triggers.
+non-rule functions including other triggers.
+
+# Tests
+
+Being able to verify the correctness policy-related logic is of paramount
+importance. As is the ability to pose ad hoc queries with partial state. To this
+end we include `@test` as supported decorator and introduce the `with-as` clause
+to assist with partial state bindings required for both adhoc queries and for
+function mocking.
+
+Note: the following is under review and not yet reflected in the grammar. 
+
+```
+@rule function user_owned_action(auth, resource) {
+  result = resource.matches('documents/{owner}/**')
+  return (result.owner == auth.uid
+       || resource.owner in user_groups(auth.principal));
+}
+
+@extern function user_groups(user);
+
+@test function group_check() {
+   function mock_user_groups(user) { … }
```

Comment:
For now this could be defined outside of the test function? Otherwise, I think this will require more changes to the existing grammar.

Reply:
Certainly, you're quite right. We don't permit nested function declarations. I will update the example.